File: src/http-client-mock/RequestBehaviorBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Json;
using JetBrains.Annotations;

namespace HttpClientMock
{
    public class RequestBehaviorBuilder
    {
        private readonly Func<string, bool> urlMatcher;
        private readonly HttpMethod method;

        private HttpStatusCode statusCode;
        private IDictionary<string, string> headers;
        private Uri location;
        private Func<HttpRequestMessage, HttpContent> createContent;
        private readonly List<RequestCapture> captures = new List<RequestCapture>();
        private Func<HttpRequestMessage, bool> matcher;

        public RequestBehaviorBuilder(Func<string, bool> urlMatcher, HttpMethod method)
        {
            this.method = method;
            this.urlMatcher = urlMatcher;
            createContent = r => JsonContent.Create(string.Empty);
            matcher = r => MatchRespond.Match<object>(r, urlMatcher, method, (capture, model) => true, captures);
        }

        [PublicAPI]
        public RequestBehaviorBuilder MatchRequest<TR>(TR schema = default, Func<RequestCapture, TR, bool> matchFunc = null)
        {
            return MatchRequest(matchFunc);
        }
        
        [PublicAPI]
        public RequestBehaviorBuilder MatchRequest(Func<RequestCapture, object, bool> matchFunc = null)
        {
            return MatchRequest<object>(matchFunc);
        }
        
        [PublicAPI]
        public RequestBehaviorBuilder MatchRequest<TR>(Func<RequestCapture, TR, bool> matchFunc = null)
        {
            matcher = r => MatchRespond.Match(r, urlMatcher, method, matchFunc?? ((capture, model) =>true) , captures);
            return this;
        }
        
        [PublicAPI]
        public Func<RequestCapture> Capture()
        {
            return () => captures.LastOrDefault();
        }
        
        [PublicAPI]
        public Func<(RequestCapture requestCapture, TR model)> Capture<TR>(TR schema = default)
        {
            return () =>
            {
                var requestCapture = captures.LastOrDefault();
                TR model = default;
                if (requestCapture != null)
                {
                    model = requestCapture.Model<TR>();
                }
                return (requestCapture, model);
            };
        }

        [PublicAPI]
        public Func<IEnumerable<RequestCapture>> CaptureAll()
        {
            return () => captures.ToArray();
        }
        
        [PublicAPI]
        public Func<IEnumerable<(RequestCapture requestCapture, TR model)>> CaptureAll<TR>(TR schema = default)
        {
            return () => captures.Select(c => (c, c.Model<TR>())).ToArray();
        }
        
        [PublicAPI]
        public RequestBehaviorBuilder Respond(HttpStatusCode httpStatusCode)
        {
            statusCode = httpStatusCode;
            return this;
        }
        [PublicAPI]
        public RequestBehaviorBuilder Respond(HttpStatusCode httpStatusCode, object content, Uri location = null)
        {
            statusCode = httpStatusCode;
            createContent = r => JsonContent.Create(content);
            this.location = location;
            return this;
        }
        [PublicAPI]
        public RequestBehaviorBuilder RespondContent(HttpStatusCode httpStatusCode, Func<HttpRequestMessage,HttpContent> contentFn, Uri location = null)
        {
            statusCode = httpStatusCode;
            createContent = contentFn;
            this.location = location;
            return this;
        }
        [PublicAPI]
        public RequestBehaviorBuilder RespondHeaders(dynamic headers)
        {
            this.headers = Dyn2Dict((object)headers);
            return this;
        }
        
        internal (Func<HttpRequestMessage, bool> Matcher, Func<HttpRequestMessage, HttpResponseMessage> Responder) Build()
        {
            return (
                Matcher: r => matcher(r),
                Responder: r => MatchRespond.Respond(r, createContent, statusCode, headers, location)
            );
        }

        private static Dictionary<string, string> Dyn2Dict(object dynObj)
        {
            return dynObj.GetType().GetProperties().ToDictionary(prop => prop.Name, 
                prop => prop.GetValue(dynObj, null).ToString());
        }
    }
}

File: src/fs-http-client-mock/RequestBehaviorBuilder.fs
namespace HttpClientMock

open System
open System.Collections.Generic
open System.Net
open System.Net.Http
open System.Net.Http.Json
open System.Runtime.InteropServices
open JetBrains.Annotations
type RequestBehaviorBuilder internal (urlMatcher: string -> bool, method: HttpMethod) =
    let Obj2Dict( dynObj: Object): Map<string, string> =
        dynObj.GetType().GetProperties()
            |> Seq.map (fun p -> (p.Name, p.GetValue(dynObj, null).ToString() ))
            |> Map.ofSeq
    
    let mutable statusCode: HttpStatusCode = HttpStatusCode.NotFound
    let mutable headerLocation: Uri option = None
    let mutable responseHeaders: Map<string, string> = Map.empty
    let mutable createContent: HttpRequestMessage -> HttpContent = fun _ -> JsonContent.Create("") :> HttpContent
    
    let mutable captures = List.empty 
    
    let mutable matcher: HttpRequestMessage -> bool =
        fun r ->
            match MatchResponse.Match<Object>(r, urlMatcher, method, (fun _ _ -> true)) with
                | Some(c) ->
                    captures <- captures @ [c]
                    true
                | None -> false

    [<PublicAPI>]
    member this.MatchRequest<'TR>(matchFunc: Func<RequestCapture, 'TR, bool>) =
        let mm: RequestCapture -> 'TR option -> bool = fun r tr ->
            match tr with
                | Some ttr -> matchFunc.Invoke(r, ttr)
                | None -> matchFunc.Invoke(r, Unchecked.defaultof<'TR> )

        matcher <- (fun r ->
            match MatchResponse.Match<'TR>(r, urlMatcher, method, mm) with
                | Some(c) ->
                    captures <- captures @ [c]
                    true
                | None -> false
            )
        this

    /// this is only for csharp use since there's no type declaration for anonymous type
    [<PublicAPI>]
    member this.MatchRequest<'TR>(matchFunc: Func<RequestCapture, 'TR, bool>, _: 'TR) =        
        this.MatchRequest<'TR>(matchFunc)
    
    [<PublicAPI>]
    member this.Capture() =
        (fun () -> captures |> Seq.tryLast)
            |> FuncUtils.ignoreOption

    [<PublicAPI>]
    member this.Capture<'TR>() =
        fun () ->            
            match captures |> Seq.tryLast with
                | Some r -> Some (r, r.Model<'TR>())
                | None -> None
        |> FuncUtils.ignoreTupleOption
        
    
    [<PublicAPI>]
    member this.Capture<'TR>(_schema: 'TR) =
        this.Capture<'TR>()
        
    [<PublicAPI>]
    member this.CaptureAll() =
        captures |> Array.ofSeq
    

    [<PublicAPI>]
    member this.CaptureAll<'TR>(?_schema: 'TR) =
        fun () -> captures |> Seq.map(fun c -> (c, c.Model<'TR>())) |> Array.ofSeq;
            
    [<PublicAPI>]
    member this.Respond(status: HttpStatusCode,
                        contentFn: Func<HttpRequestMessage, HttpContent>,
                        [<Optional>]?headers: struct(string*string) seq,
                        [<Optional>]?location: Uri) =        
        statusCode <- status
        createContent <- contentFn.Invoke
        responseHeaders <- (defaultArg headers Seq.empty )
                           |> Seq.map (fun h -> h.ToTuple())
                           |> Map.ofSeq
        headerLocation <- location
        this
        
    [<PublicAPI>]
    member this.Respond(status: HttpStatusCode,
                        [<Optional>]?content: Object,
                        [<Optional>]?headers: struct(string*string) seq,
                        [<Optional>]?location: Uri) =
        let createContent = fun _ -> match content with
                                        | None -> JsonContent.Create("") :> HttpContent
                                        | Some c -> JsonContent.Create(c) :> HttpContent

        this.Respond(status, contentFn = Func<HttpRequestMessage,HttpContent>(createContent),
                     ?headers = headers, ?location = location)
        
    member internal this.Build() : Behavior =
        {
            Matcher = matcher
            Responder = fun r -> MatchResponse.Respond(r, createContent, statusCode, responseHeaders, headerLocation)
        }
